pkg/daemon/update.go

```diff
@@ -556,6 +587,122 @@ func (dn *Daemon) updateKernelArguments(oldConfig, newConfig *mcfgv1.MachineConf
 	return exec.Command("rpm-ostree", args...).Run()
 }
 
+// MountOSContainer mounts the container and returns the mountpoint
+func (dn *Daemon) MountOSContainer(container string) (mnt, containerName string, err error) {
+
+	var authArgs []string
+	if _, err = os.Stat(kubeletAuthFile); err == nil {
+		authArgs = append(authArgs, "--authfile", kubeletAuthFile)
+	}
+	// Pull the image
+	args := []string{"pull", "-q"}
+	args = append(args, authArgs...)
+	args = append(args, container)
+	pivotutils.RunExt(false, numRetriesNetCommands, "podman", args...)
+
+	containerName = "mcd-" + string(uuid.NewUUID())
+	// `podman mount` wants a container, so let's create a dummy one, but not run it
+	var cidBuf []byte
+	cidBuf, err = runGetOut("podman", "create", "--net=none", "--annotation=org.openshift.machineconfigoperator.pivot=true", "--name", containerName, container)
+	if err != nil {
+		return
+	}
+
+	cid := strings.TrimSpace(string(cidBuf))
+	// Use the container ID to find its mount point
+	mntBuf, err := runGetOut("podman", "mount", cid)
+	if err != nil {
+		return
+	}
+	mnt = strings.TrimSpace(string(mntBuf))
+	return
+}
+
+// switchKernel updates kernel on host with the kernelType specified in MachineConfig.
+// Right now it supports default (traditional) and realtime kernel
+func (dn *Daemon) switchKernel(oldConfig, newConfig *mcfgv1.MachineConfig) error {
+	// We support Kernel update only on RHCOS nodes
+	if dn.OperatingSystem != machineConfigDaemonOSRHCOS {
+		return fmt.Errorf("Updating kernel on non-RHCOS nodes is not supported")
+	}
+
+	dn.logSystem("ImageURl  %v - %v ", oldConfig.Spec.OSImageURL, newConfig.Spec.OSImageURL)
+	defaultKernel := []string{"kernel", "kernel-core", "kernel-modules", "kernel-modules-extra"}
+	dn.logSystem("Initiating switch from kernel %v to %v", ValidKernelType(oldConfig.Spec.KernelType), ValidKernelType(newConfig.Spec.KernelType))
+	if ValidKernelType(oldConfig.Spec.KernelType) == realtimeKernelType && ValidKernelType(newConfig.Spec.KernelType) == defaultKernelType {
```

Comment:
Yeah, they are valid KernelType at this point. Since KernelType is string, for unset kernelType it contains empty (""). For additional safety. I am using this function to make `default `and empty ("") considered as `default` before comparison